File: App/Resources/AlgorithmDetails/hybridcombsort.bundle/hybridcombsort.go
package main

import (
  "fmt"
)

func insertionSort(arr []int) {
  n := len(arr)
  for i := 1; i < n; i++ {
    key := arr[i]
    j := i - 1
    for j >= 0 && arr[j] > key {
      arr[j+1] = arr[j]
      j = j - 1
    }
    arr[j+1] = key
  }
}

func sort(arr []int) []int {
  n := len(arr)
  shrink := 1.3
  gap := n
  sorted := false
  threshold := n / 32
  if threshold > 8 {
    threshold = 8
  }
  for !sorted {
    gap = int(float64(gap) / shrink)
    if gap <= 1 {
      sorted = true
      gap = 1
    }
    for i := 0; i < n-gap; i++ {
      if gap <= threshold {
        insertionSort(arr)
        return arr
      }
      sm := gap + i
      if arr[i] > arr[sm] {
        arr[i], arr[sm] = arr[sm], arr[i]
        sorted = false
      }
    }
  }
  return arr
}

func main() {
  array := []int{0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56}
  fmt.Println(sort(array))
}

File: App/Resources/AlgorithmDetails/hybridcombsort/hybridcombsort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

void insertionSort(int arr[], int n) {
  for (int i = 1; i < n; i++) {
    int key = arr[i];
    int j = i - 1;
    while (j >= 0 && arr[j] > key) {
      arr[j + 1] = arr[j];
      j = j - 1;
    }
    arr[j + 1] = key;
  }
}

void sort(int arr[], int n) {
  float shrink = 1.3f;
  int gap = n;
  int sorted = 0;
  int threshold = n / 32;
  if (threshold > 8) {
    threshold = 8;
  }
  while (!sorted) {
    gap = (int)((float)gap / shrink);
    if (gap <= 1) {
      sorted = 1;
      gap = 1;
    }
    for (int i = 0; i < n - gap; i++) {
      if (gap <= threshold) {
        insertionSort(arr, n);
        return;
      }
      int sm = gap + i;
      if (arr[i] > arr[sm]) {
        swap(&arr[i], &arr[sm]);
        sorted = 0;
      }
    }
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}
